core/store/src/validate.rs

```diff
@@ -0,0 +1,260 @@
+    }
+}
+
+fn block_of_chunk_exists(
+    store: &Store,
+    _config: &GenesisConfig,
+    _key: &[u8],
+    value: &[u8],
+) -> Result<(), String> {
+    let shard_chunk = ShardChunk::try_from_slice(value).unwrap();
+    let height = shard_chunk.header.height_included;
+    if height == 0 {
+        // TODO discuss - non-accepted chunk found?
+        return Ok(());
+    }
+    match store.get_ser::<HashMap<EpochId, HashSet<CryptoHash>>>(
+        ColBlockPerHeight,
+        &index_to_bytes(height),
+    ) {
+        Ok(Some(map)) => {
+            for (_, set) in map {
+                for block_hash in set {
+                    match store.get_ser::<Block>(ColBlock, block_hash.as_ref()) {
+                        Ok(Some(block)) => {
+                            if block.chunks.contains(&shard_chunk.header) {
+                                // Block for ShardChunk is found
+                                return Ok(());
+                            }
+                        }
+                        _ => {}
+                    }
+                }
+            }
+            err!(format!("No Block on height {:?} accepts ShardChunk {:?}", height, shard_chunk))
+        }
+        Ok(None) => err!(format!("Map not found on height {:?}", height)),
+        Err(e) => err!(format!("Can't get Map from storage on height {:?}, {:?}", height, e)),
+    }
+}
+
+fn block_height_cmp_tail(
+    store: &Store,
+    config: &GenesisConfig,
+    key: &[u8],
+    value: &[u8],
+) -> Result<(), String> {
+    let block_hash = CryptoHash::try_from(key.as_ref()).unwrap();
+    let tail = match store.get_ser::<BlockHeight>(ColBlockMisc, TAIL_KEY) {
+        Ok(Some(tail)) => tail,
+        Ok(None) => config.genesis_height,
+        Err(_) => return err!("Can't get Tail from storage"),
+    };
+    match Block::try_from_slice(value) {
+        Ok(block) => {
+            if block.header.inner_lite.height < tail
+                && block.header.inner_lite.height != config.genesis_height
+            {
+                err!(format!(
+                    "Invalid block height stored: {:?}, tail: {:?}",
+                    block.header.inner_lite.height, tail
+                ))
+            } else {
+                Ok(())
+            }
+        }
+        Err(e) => {
+            err!(format!("Can't get Block from storage, {:?}, {:?}", block_hash.to_string(), e))
+        }
+    }
+}
+
+//
+// All validations end here
+
+#[derive(Debug)]
+pub struct ErrorMessage {
+    pub col: DBCol,
+    pub msg: String,
+}
+
+impl ErrorMessage {
+    fn new(col: DBCol, msg: String) -> Self {
+        Self { col, msg }
+    }
+}
+
+#[derive(Default, BorshDeserialize)]
+pub struct StoreValidator {
+    #[borsh_skip]
+    pub errors: Vec<ErrorMessage>,
+    tests: u64,
+}
+
+impl StoreValidator {
+    pub fn is_failed(&self) -> bool {
+        self.tests == 0 || self.errors.len() > 0
+    }
+    pub fn failed(&self) -> u64 {
```

Comment:
maybe `num_failed` is a better name?

Reply:
I think `failed` is quite understandable

Reply:
@Kouprin I agree with @bowenwang1996 that this is not understandable from the call-site point of view:

```rust
println!("Errors found: {}", Red.bold().paint(store_validator.failed().to_string()));
```

Reply:
Two people said it's confusing - that's a strong sign to change.
++